session/txn.go

```diff
@@ -83,8 +84,27 @@ func (txn *LazyTxn) CacheTableInfo(id int64, info *model.TableInfo) {
 func (txn *LazyTxn) init() {
 	txn.mutations = make(map[int64]*binlog.TableMutation)
 	txn.mu.Lock()
-	txn.mu.TxnInfo.State = txninfo.TxnIdle
-	txn.mu.Unlock()
+	defer txn.mu.Unlock()
+	txn.mu.TxnInfo = txninfo.TxnInfo{}
+}
+
+// call this under lock!
+func (txn *LazyTxn) updateState(state txninfo.TxnRunningState) {
+	if txn.mu.TxnInfo.State != state {
+		lastState := txn.mu.TxnInfo.State
+		lastStateChangeTime := txn.mu.TxnInfo.LastStateChangeTime
+		txn.mu.TxnInfo.State = state
+		txn.mu.TxnInfo.LastStateChangeTime = time.Now()
+		if !lastStateChangeTime.IsZero() {
+			hasLockLbl := "false"
+			if !txn.mu.TxnInfo.BlockStartTime.IsZero() {
+				hasLockLbl = "true"
+			}
+			metrics.TxnDurationHistogram.WithLabelValues(txninfo.StateLabel(lastState), hasLockLbl).Observe(time.Since(lastStateChangeTime).Seconds())
+			metrics.TxnStatusGauge.WithLabelValues(txninfo.StateLabel(lastState)).Dec()
+		}
+		metrics.TxnStatusGauge.WithLabelValues(txninfo.StateLabel(state)).Inc()
```

Comment:
Can remove line98~line106 out of txn.mu.lock?
another question, who will access the "txn" except current session?

Reply:
> Can remove line98~line106 out of txn.mu.lock?
Most places call `updateStatus` use `defer` to unlock `txn.mu`, and I don't observe obvious regression here, so IMO though it is possible, it is not necessary.

> who will access the "txn" except current session?
For example, the user can access `tidb_trx` at the same time of committing a transaction.